File: frameworks/jax/kv_cache_manager.cc
// Copyright 2026 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "frameworks/jax/kv_cache_manager.h"

#include <cstdint>
#include <utility>
#include <vector>

#include "core/utils.h"
#include "frameworks/jax/utils.h"

namespace nb = nanobind;

namespace tpu_raiden {
namespace kv_cache {
namespace jax {

namespace {
UnpackedCache UnpackAndMove(nanobind::list device_arrays) {
  auto layer_buffers = tpu_raiden::jax::UnpackJaxArrays(device_arrays);
  return {std::move(layer_buffers), std::move(device_arrays)};
}
}  // namespace

KVCacheManager::KVCacheManager(
    nb::list device_arrays, int block_size, std::optional<int> local_port,
    std::optional<int> host_blocks_to_allocate,
    std::optional<std::vector<uintptr_t>> external_host_ptrs,
    bool unsafe_skip_buffer_lock, int parallelism)
    // NOTE: To achieve zero-copy initialization while remaining robust against
    // unspecified C++ function/constructor argument evaluation order (Clang
    // typically evaluates left-to-right, GCC evaluates right-to-left), we
    // enforce sequencing through a helper function `UnpackAndMove`. The
    // function call boundary acts as a strict sequencing barrier, guaranteeing
    // that `UnpackJaxArrays` is fully evaluated before the Python list handle
    // is moved into ownership, preventing use-after-move segfaults.
    : KVCacheManager(UnpackAndMove(std::move(device_arrays)), block_size,
                     local_port, host_blocks_to_allocate, external_host_ptrs,
                     unsafe_skip_buffer_lock, parallelism) {}

KVCacheManager::KVCacheManager(
    UnpackedCache&& cache, int block_size, std::optional<int> local_port,
    std::optional<int> host_blocks_to_allocate,
    std::optional<std::vector<uintptr_t>> external_host_ptrs,
    bool unsafe_skip_buffer_lock, int parallelism)
    : KVCacheManagerBase(
          cache.layer_buffers, block_size, local_port, host_blocks_to_allocate,
          tpu_raiden::CastExternalPointers(external_host_ptrs),
          unsafe_skip_buffer_lock, parallelism,
          tpu_raiden::CreateHostMemoryAllocator(
              cache.layer_buffers.empty() || cache.layer_buffers[0].empty()
                  ? nullptr
                  : cache.layer_buffers[0][0]->device()->client())),
      device_arrays_(std::move(cache.device_arrays)) {}

KVCacheManager::~KVCacheManager() = default;

}  // namespace jax
}  // namespace kv_cache
}  // namespace tpu_raiden

File: api/torch/kv_cache_manager.py
# Copyright 2026 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Copyright 2026 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""High-performance PyTorch KV Cache Manager (repurposed as Raw Transfer)."""

from frameworks.torch import _torch_raw_transfer as _impl

# Expose the core stateless raw transfer API classes and functions
RawHostBuffer = _impl.RawHostBuffer
PjRtCopyFuture = _impl.PjRtCopyFuture
PreparedTorchRawTransfer = _impl.PreparedTorchRawTransfer

await_all = _impl.await_all
is_ready = _impl.is_ready

transfer_d2h_async = _impl.transfer_d2h_async
transfer_h2d_async = _impl.transfer_h2d_async
transfer_d2h = _impl.transfer_d2h
transfer_h2d = _impl.transfer_h2d

transfer_d2h_batch_async = _impl.transfer_d2h_batch_async
transfer_h2d_batch_async = _impl.transfer_h2d_batch_async
transfer_d2h_batch = _impl.transfer_d2h_batch
transfer_h2d_batch = _impl.transfer_h2d_batch
